### PR TITLE
Book 3 Edits Chapters 10-13

### PR DESCRIPTION
**Pull request by trevordblack**
I'm genuinely amazed at how much I got done in the last month.
But, anyway, here are the edits for the last bit of book 3.
There were necessary bits of source code change that I wasn't able to compile to confirm that they work.

We'll need to go through and redo the renders to confirm functionality on everything, but we already knew that.

I still need to go back through this again to iron out a collection of small little things, but this is the bulk of what I'm going to be doing until the new year.

I'd also like Peter to go through and confirm that the reading makes sense and still is in his "voice". All for another day though.
Onto Eggnog!

### CODE REVIEW

File: books/RayTracingTheRestOfYourLife.html [hunk 1]
```diff
@@ -3335,9 +3465,9 @@
 The Rest of Your Life
 ====================================================================================================
 
-The purpose of this book was to show the details of dotting all the i’s of the math on one way of
-organizing a physically based renderer’s sampling approach. Now you can explore a lot of different
-potential paths.
+The purpose of this book was to walk through all of the little details (dotting all the i's and
+crossing all of the t's) necessary when organizing a physically based renderer’s sampling approach.
+You should now be able to take all of this detail and explore a lot of different potential paths.
```
Comment (hollasch) [inline books/RayTracingTheRestOfYourLife.html:3470]: Nice rewrite.

**Review by hollasch — APPROVED** [review]
Again, nice job.

File: books/RayTracingTheRestOfYourLife.html [hunk 2]
```diff
@@ -2403,50 +2402,41 @@
...
+$\cos(\theta)/\pi$. In a couple of our examples we generated the random direction using a different
+distribution than the distribution of the PDF. We covered this quite a lot in the
+chapter Playing with Importance Sampling. In general, if we know the distribution of our random
+directions, we should use a PDF with the same distribution, this will lead to the fastest
+convergence. With that in mind, we can create a `pdf` class that is responsible both generating
+random directions and determining the value of the PDF. From all of this, any `pdf` class should be
+responsible for,
 
+1. Generate a random direction weighted by the internal PDF distribution
+2. What is the PDF distribution value for that direction?
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:2422]: I don't quite follow? What do you mean for these comments?

File: books/RayTracingTheRestOfYourLife.html [hunk 3]
```diff
@@ -2550,27 +2535,32 @@
...
 <div class='together'>
-This assumes two as-yet not implemented functions in the `hittable` class. To avoid having to add
-instrumentation to all `hittable` subclasses, we’ll add two dummy functions to the `hittable` class:
+If we want to sample the light, we will need `hittable` to answer some queries that it doesn’t yet
+have an interface for. The above code assumes the existence of two as-of-yet unimplemented functions
+in the `hittable` class: `pdf_value()` and `random()`. We'll need to add these functions for the
+program to compile. We could go through all of the `hittable` subclasses and add these functions,
+but that would be a hassle, so we’ll just add two trivial functions to the `hittable` base class.
+This breaks our previously pure abstract implementation, but it saves work. Feel free to write these
+functions through to subclasses if you want a purely abstract `hittable` interface class.
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:2563]: I waffled on this one. Ultimately, I didn't (and don't) think it's enough to break the implicit abstract implementation without a mention. I'm having a tremendously difficult time articulating _why_ I feel that way.

But, to the simplest extent, I think that the break in pure abstract is noteworthy, and I tried to mention anything I thought was noteworthy.

I think it also cuts through the monotony of all of the code copying in this chapter.

I don't know.

File: books/RayTracingTheRestOfYourLife.html [hunk 4]
```diff
@@ -2756,53 +2810,60 @@
...
-under doors or whatever else you think might be bright. But most programs branch, and send one or
-more terminal rays to lights explicitly, and one according to the reflective distribution of the
-surface. This could be a time you want faster convergence on more restricted scenes and add shadow
-rays; that’s a personal design preference.
+We won't write any code in this chapter. We’re at a crossroads and we need to make some
+architectural decisions. The mixture-density approach is an alternative to having more traditional
+shadow rays, which are rays that check to see if an intersection is lit by the lights in the scene.
+The mixture-density approach is something that I personally prefer, because in addition to lights
+you can sample windows or bright cracks under doors or whatever else you think might be bright --
+or important. But you'll see shadow rays in most professional path tracers. Typically they'll have
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:2828]: "still see" implies that it's uncommon. Peter's preference for mixture density is the uncommon one. Shadow rays are pretty much universal at this point

[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:2828]: mmm

[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:2828]: Made change. You're right.

File: books/RayTracingTheRestOfYourLife.html [hunk 5]
```diff
@@ -2756,53 +2810,60 @@
...
-The specular rays (glass and metal) are no longer supported. The math would work out if we just made
-their scattering function a delta function. But that would be floating point disaster. We could
-either separate out specular reflections, or have surface roughness never be zero and have
-almost-mirrors that look perfectly smooth but don’t generate NaNs. I don’t have an opinion on which
-way to do it (I have tried both and they both have their advantages), but we have smooth metal and
-glass code anyway, so I add perfect specular surfaces that do not do explicit f()/p() calculations.
+We've accidentally broken the specular rays (glass and metal) and they are no longer supported.
+The math would continue to work out if we just made their scattering function a delta function. But
+that would lead to all kinds of floating point disasters. We could either make specular reflections
+a special case that skips $f()/p()$, or we could set surface roughness to a really small -- but
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:2851]: The important bit here is that the surface roughness is a number. That number becoming zero is what creates NaNs. We need that number to be small, but nonzero.
Qualifying the surface roughness as "smooth" obscures that the wonkiness of needing a number

File: books/RayTracingTheRestOfYourLife.html [hunk 6]
```diff
@@ -2815,43 +2876,40 @@
...
-        bool is_specular;
-        color attenuation;
-        shared_ptr<pdf> pdf_ptr;
-    };
+        public:
+          color attenuation;
+          shared_ptr<pdf> pdf_ptr;
+          bool skip_pdf;
+          ray skip_pdf_ray;
+      };
```
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:2906]: Except for the spacing. Whoops.

File: books/RayTracingTheRestOfYourLife.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (hollasch) [reply books/RayTracingTheRestOfYourLife.html:2422]: For the second comment, you begin stating that a `pdf` class should be responsible for two things, but then the second thing is a question. Re-reading it, are you asking the question in item 2, or were you trying to state that the `pdf` class should have a method that answers that question? Is this more accurate?

> With that in mind, we can create a `pdf` class that is responsible both generating random directions and determining the value of the PDF. Thus, any `pdf`-derived class will need to implement two methods:
>
> 1. Return a random direction weighted by the internal PDF distribution, and
> 2. Return the corresponding PDF distribution value in that direction

File: books/RayTracingTheRestOfYourLife.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (hollasch) [reply books/RayTracingTheRestOfYourLife.html:2563]: ¯\_(ツ)_/¯ Your call.

File: books/RayTracingTheRestOfYourLife.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (hollasch) [reply books/RayTracingTheRestOfYourLife.html:2851]: Makes sense.

File: books/RayTracingTheRestOfYourLife.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:2422]: The latter

File: books/RayTracingTheRestOfYourLife.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (trevordblack) [reply books/RayTracingTheRestOfYourLife.html:2563]: keeping it in

**Review by hollasch — APPROVED** [review]

**Pull request merged**